### PR TITLE
Create and use new AP_HAL::PWMSource

### PR DESCRIPTION
**Pull request by peterbarker**
This factors out some common code from several classes into the HAL.

A third candidate, `AP_RPM_Pin` is not shifted as it does some filtering on the input pin which may actually be a good idea to have generally - but wasn't in the other drivers.  Similarly for the FuelFlow driver (but not the FuelLevel driver!)

This work was done because I was looking at adding *another* very similar implementation and thought we should do better.

### CODE REVIEW

File: libraries/AP_HAL/GPIO.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        return interrupt_attached;
+    }
+
+    if (interrupt_attached) {
+        if (!hal.gpio->detach_interrupt(_pin)) {
+            GCS_SEND_TEXT(MAV_SEVERITY_WARNING,
+                          "%s: Failed to detach interrupt from %d",
+                          subsystem,
+                          _pin);
+            // memcpy(sand, head, sizeof(sand))
```
Comment (tridge) [inline libraries/AP_HAL/GPIO.cpp:25]: remove debug

File: libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp [hunk 2]
```diff
@@ -116,35 +79,19 @@ bool AP_RangeFinder_HC_SR04::detect(AP_RangeFinder_Params &_params)
...
 /*
   update distance_cm
  */
 void AP_RangeFinder_HC_SR04::update(void)
 {
     // check if pin has changed and configure interrupt handlers if required:
-    check_pins();
-
-    if (!echo_pin || ! trigger_pin || echo_pin == -1 || trigger_pin == -1) {
+    if (!check_pins()) {
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp:88]: is this a behaviour change for pin values of -1?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL/GPIO.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_HAL/GPIO.cpp:25]: done

File: libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp:88]: Checks have moved.

**Pull request merged**